accepted/global-json-updates.md

```diff
@@ -0,0 +1,369 @@
+  * Due to Rule #2, there is no roll forward from preview to stable if they are both on the machine.
+    * The impact of this is mitigated by work we did in .NET Core 3.0 Preview 3 to replace earlier patch versions.
+* User wants to always use a specific SDK patch version, and receive an error if it is not on the machine.
+  * The user cannot currently do this.
+* User wants to always run the latest patch of a particular SDK feature band.
+  * Due to Rule #2, the only way to do this is to remove the previous version.
+    * The impact of this is mitigated by work we did in .NET Core 3.0 Preview 3 to replace earlier patch versions.
+* User includes a feature that requires an SDK that corresponds to a SDK feature band or higher, and wants to run the best available tools.
+  * There is no mechanism to require at least a specific version, but otherwise use the latest, or the latest in a range.
+  * Upgrading to a newer SDK will provide no benefit to projects tied to an earlier version.
+* User wants to know what version of the SDK is being used.
+  * `global.json` is a subtle, hidden and easily forgotten marker – programmers may not understand the downsides of leaving a global.json that was suggested in a blog post in place – or they may forget it.
+* User rarely wants the risk of running the unsupported scenario of multiple SDK versions in a solution build.
+  * There is nothing that encourages users to place `global.json` in the repository root or the same level as the solution. Using different versions of the SDK for different projects in the solution is not a supported scenario.
+* Hosted CI, including Azure DevOps and Kudu, need to provide a small number of SDKs.
+  * If there is no patch of the major/minor/feature version (xyz in `x.y.znn`) present, the build fails, causing friction between user and host needs.
+
+## Proposal
+
+This proposal parallels [runtime roll-forward behavior as initially described](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/roll-forward-on-no-candidate-fx.md)} and [recent improvements](https://github.com/dotnet/designs/blob/master/accepted/runtime-binding.md). All of the features proposed here may not be included in Phase 1. It will include at least `usePreview` and `"rollforward": "disable"`
+
+The runtime options allow for three basic approaches: find nearest, find latest and find exact. 
+
+## Details
+
+* Runtime: From the [recent improvements](https://github.com/dotnet/designs/blob/master/accepted/runtime-binding.md) document" _In all cases except `disable` the highest available patch version is selected._"
+
+The `global.json` schema to offers support for all of the identified scenarios that can be fixed without altering how we find `global.json`:
+
+* `sdk`
+  * `version`
+    * The lowest SDK the resolver would select - any SDKs with lower versions are ignored.
+    * Identical to today - wildcards are not supported.
+    * Default: empty
+  * `usePreview`
```

Comment:
In SemVer the correct term is pre-release. In the framework roll forward we even have an environment variable which uses the term "pre-release" for this.

I think this needs a much more detailed description of the desired behavior. Again in framework roll forward we actually spent a lot of time on this: https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/framework-version-resolution.md#pre-release-versions

In short:
* If the `version` is a release version, should we try to fulfill the request with release first and only then fallback to pre-release? (even is `usePreview` is true).
* What happens if `version` is pre-release but `usePreview=false`
* If `usePreview=false` and we can't find any version and are about to fail - should we consider pre-release versions in this case to avoid failing?

In the below text you always state that we pick the latest patch in the given range (except for disable). We should be explicit what it means for pre-release versions. Do we always pick latest pre-release for a given patch? We actually went back and forth on this for frameworks and ended up having a somewhat in-between solution:
* If the `version` is pre-release, and there is an exact match, use it - don't roll to latest patch/pre-release - otherwise it's really hard to test several different pre-release versions side-by-side. Also we tend to treat pre-release versions as breaking (P4 versus P8 - there might be breaking changes there), so it's better to pick "closest" in this case, to avoid breaking people. I know that we mostly mitigate this by uninstalling other previews, but still - even for internal testing this is important.

Reply:
The setting is `allowPrerelease`.  This part is out of date.

Reply:
To answer questions:

> If the version is a release version, should we try to fulfill the request with release first and only then fallback to pre-release? (even is usePreview is true).

If I specify `version = 2.2.300, rollForward = latestMajor, allowPrerelease = true` and have `2.2.505` and `3.0.100-preview7` installed, I would be quite surprised if it prefers 2.2.505 over 3.0.100-preview7.  I don't think the version should have semantic meaning regarding whether or not prereleases are matched against (see next answer).

> What happens if version is pre-release but usePreview=false

Currently it silently allows prerelease versions.  My changes make this a warning saying we're ignoring your `usePreview` setting and proceed to allow prereleases.  Personally, I don't like either of these behaviors as to me `version` is a *minimum* bound for the roll-forward policy and shouldn't otherwise have semantic meaning.

> If usePreview=false and we can't find any version and are about to fail - should we consider pre-release versions in this case to avoid failing?

I don't think so.  If I instruct my tools to not consider prerelease versions then it should do so.

> Do we always pick latest pre-release for a given patch?

As this design is currently going to be implemented, yes.  Users should be able to explicitly use a specific prerelease patch level by using either the `patch` or `disable` policies.  Generally users don't install preview servicing releases and our testers don't test multiple preview servicing releases side by side, so I think the impact here would be low.

Reply:
OK - makes sense, I just think this should be explicitly stated - if nothing else so that there's a doc which can be compared against tests/actual behavior.